# Data Wrangling Project of Pune_India.osm

## by Sebastian P Joseph

Pune is a sprawling city in the western Indian state of Maharashtra. It was once the base of the Peshwas (prime ministers) of the Maratha Empire, which lasted from 1674 to 1818. It's known for the grand Aga Khan Palace, built in 1892 and now a memorial to Mahatma Gandhi, whose ashes are preserved in the garden. The 8th-century Pataleshwar Cave Temple is dedicated to the Hindu god Shiva. 

Moreover its my hometown , and its one of the most upcoming city in INDIA. So i felt i could take up that city for my project and explore a bit more.

When I explored the data a bit i found that most of the errors are found in the street names area and the postal codes area.

## Problems

1. Street names

2. Postal codes

## 1. Street names 

In the first part of the code i have decided to display all the street names. I have mentioned some expected words to come in the end which might come in the pune dataset. They are as follows.

expected = ["Street", "Avenue","Apartment","Ridge","Annexe","City","Society","Area","Bunk","Centre","Station","Circle",
            "Nagar","Stage","Stop","Lake","World","Plaza","Phase","Bazaar","Park","Colony","Commons","Mall","Trail",
            "Path","Road","Gate","Town","Drive","Block"]
            
I have printed all the street types as you can see in the notebook file.

In the next part I have tried to clean some street types which i could find. They are mentioned below.

mapping={" , Pune":"",
         "cross":"Cross",
        "Rd.":"Road",
        "Rd":"Road",
        "road":"Road",
        "raod":"Road",
        "apartment":"Apartment",
        "no.":"",
        "udyog":"Udyog",
        "pedestrian":"Pedestrian"
        }
        
I have tried to map them to the original words and printed the output using the variable update_street.

## 2. Postal Codes 

In the second part i have decided to display the postal codes. everything ,if there is error in it or not.

Then exploring it a bit,I found that some postal codes have spaces in between. Eg: 413 321

Then some postal codes are in the form of string.

Then some postal codes are out of the city limits. The city postal codes range is from 411001 to 411053

Then I started cleaning the postal codes and printed all the cleaned postal codes.

## 3. SQL

In this part of the code i have created the CSV files required to create the database to pass the queries.

NODES_PATH = "nodes.csv"

NODE_TAGS_PATH = "nodes_tags.csv"

WAYS_PATH = "ways.csv"

WAY_NODES_PATH = "ways_nodes.csv"

WAY_TAGS_PATH = "ways_tags.csv"

The details of the files are as follows:

nodes.csv (114MB)

nodes_tags.csv (528KB)

ways.csv (15.9MB)

ways_nodes.csv (40.7MB)

ways_tags.csv (9.27MB)

The next step is to create database. I have created a database named seb_pune.db using the database.py python file attached.

## Queries and answers

## 1. Number of nodes:

In [2]:
import sqlite3
data=sqlite3.connect("seb_pune.db")
c=data.cursor()
QUERY='''SELECT count(*)as num from nodes;'''
c.execute(QUERY)
rows=c.fetchall()
import pandas as pd    
dataframe=pd.DataFrame(rows)
print dataframe
data.close()

         0
0  1418393


## 2. Number of ways

In [4]:
import sqlite3
data=sqlite3.connect("seb_pune.db")
c=data.cursor()
QUERY='''SELECT count(*)as num from ways;'''
c.execute(QUERY)
rows=c.fetchall()
import pandas as pd    
dataframe=pd.DataFrame(rows)
print dataframe
data.close()

        0
0  270313


## 3. Number of unique users 

In [5]:
import sqlite3
data=sqlite3.connect("seb_pune.db")
c=data.cursor()
QUERY='''SELECT DISTINCT(user) from (select user from nodes UNION ALL select user from ways);'''
c.execute(QUERY)
rows=c.fetchall()
import pandas as pd    
dataframe=pd.DataFrame(rows)
print dataframe
print len(dataframe),"Unique users"
data.close()

                           0
0                   PlaneMad
1                      norky
2                  singleton
3                    Heinz_V
4                        JND
5            Rahul Maindargi
6                  katpatuka
7                  shalinins
8                 mapusdotin
9              nirvanareborn
10            Pradnya shende
11                   shiva05
12                adivik2000
13                    Mettal
14                  AdyShake
15                AbhishekAJ
16             Manas Sambhus
17                 maheshrkm
18                Navaneetha
19                  mahesham
20                     vraju
21                   aveekbh
22                   cdavila
23                 parambyte
24                    uday01
25             Indrayudh Roy
26                   fulcrum
27                    Tinkle
28   Praveen Arimbrathodiyil
29                  Oberaffe
..                       ...
661                  SUVIRON
662        prashant bhaskara
663           

## 4. Top 10 contributing users

In [6]:
import sqlite3
data=sqlite3.connect("seb_pune.db")
c=data.cursor()
QUERY='''SELECT user,count(user)  from (select user from nodes UNION ALL select user from ways)
group by user 
order by  count(user) desc 
limit 10;'''
c.execute(QUERY) 
rows=c.fetchall()
import pandas as pd    
dataframe=pd.DataFrame(rows)
print dataframe
data.close()

               0      1
0      singleton  96636
1    harishvarma  60143
2  jasvinderkaur  57694
3        sramesh  57627
4       praveeng  56788
5        shiva05  51899
6    anushapyata  49530
7   kranthikumar  47440
8        harishk  43180
9       saikumar  40332


## 5. Biggest religion

In [7]:
import sqlite3
data=sqlite3.connect("seb_pune.db")
c=data.cursor()
QUERY='''SELECT value,count(*)as num from (select value,key from nodes_tags  UNION ALL select value,key from ways_tags)
where key='religion'
group by value
order by num desc
limit 1;'''
c.execute(QUERY)
rows=c.fetchall()
import pandas as pd    
dataframe=pd.DataFrame(rows)
print dataframe
data.close()

       0    1
0  hindu  227


## 6. Top 10 appearing amenities

In [8]:
import sqlite3
data=sqlite3.connect("seb_pune.db")
c=data.cursor()
QUERY='''SELECT value,count(*)as num from (select value,key from nodes_tags  UNION ALL select value,key from ways_tags)
where key='amenity'
group by value
order by num desc
limit 10;'''
c.execute(QUERY)
rows=c.fetchall()
import pandas as pd    
dataframe=pd.DataFrame(rows)
print dataframe
data.close()

            0    1
0  restaurant  153
1        cafe   55
2    hospital   48
3   fast_food   46
4        bank   45
5      school   39
6     toilets   34
7         atm   30
8        fuel   25
9    pharmacy   24


## Explanations and Suggestions: 

From the above queries we know the following:
    
1 Total number of nodes are 1418393

2 Total number of ways are 270313

3 There are 691 unique users

4 Singleton contributes the most 

5 Biggest religion in Pune is HINDU

6 Restaurant tops the amenities list in PUNE
 
 Thinking about these values, I’m reminded of “gamification” as a motivating force for contribution. In the context of the OpenStreetMap, if user data were more prominently displayed, perhaps others would take an initiative in submitting more edits to the map. And, if everyone sees that only a handful of power users are creating more than 90% a of given map, that might spur the creation of more efficient bots, especially if certain gamification elements were present, such as rewards, badges, or a leaderboard. But it is easy to give the suggestions but, there are also equal dis advantages to it.The authenticity of the map might be under risk. I guess that might be a  problem.

# Conclusion

From the above mentioned explanation I think I have possibly done my part to clean the dataset and I have answered the queries needed. After this review of the data it’s obvious that the Pune area is incomplete, though I believe it has been well cleaned for the purposes of this exercise. It interests me to notice a fair amount of GPS data makes it into OpenStreetMap.org on account of users’ efforts, whether by scripting a map editing bot or otherwise. 